pkg/agent/cniserver/server_windows.go

```diff
@@ -76,14 +76,17 @@ func getInfraContainer(containerID, netNS string) string {
 }
 
 // getInfraContainer returns the infra (sandbox) container ID of a Pod.
-// On Windows, kubelet sends two kinds of CNI ADD requests for each Pod:
+// On Windows, dockershim/containerd sends two sequentiral CNI ADD requests when a new pod is created...
```

Comment:
@jayunit100 what we observed with dockershim was not only two sequetial requests but multiple requests of two kinds:
```
I0716 01:28:33.949849     508 server.go:375] Received CmdAdd request cni_args:<container_id:"d636c4374967a6f90d9b802d9bf24cb4f060e716b0ec8165fcc59be2a59daf28" netns:"none" ifname:"eth0" args:"IgnoreUnknown=1;K8S_POD_NAMESPACE=network-policy-1573;K8S_POD_NAME=server-g4kvm;K8S_POD_INFRA_CONTAINER_ID=d636c4374967a6f90d9b802d9bf24cb4f060e716b0ec8165fcc59be2a59daf28" path:"/opt/cni/bin" network_configuration:"{\"capabilities\":{\"dns\":true},\"cniVersion\":\"0.3.0\",\"ipam\":{\"type\":\"host-local\"},\"name\":\"antrea\",\"runtimeConfig\":{\"dns\":{\"servers\":[\"10.96.0.10\"],\"searches\":[\"network-policy-1573.svc.cluster.local\",\"svc.cluster.local\",\"cluster.local\"],\"options\":[\"ndots:5\"]}},\"type\":\"antrea\"}" >
I0716 01:28:36.470742     508 server.go:375] Received CmdAdd request cni_args:<container_id:"d636c4374967a6f90d9b802d9bf24cb4f060e716b0ec8165fcc59be2a59daf28" netns:"none" ifname:"eth0" args:"IgnoreUnknown=1;K8S_POD_NAMESPACE=network-policy-1573;K8S_POD_NAME=server-g4kvm;K8S_POD_INFRA_CONTAINER_ID=d636c4374967a6f90d9b802d9bf24cb4f060e716b0ec8165fcc59be2a59daf28" path:"/opt/cni/bin" network_configuration:"{\"capabilities\":{\"dns\":true},\"cniVersion\":\"0.3.0\",\"ipam\":{\"type\":\"host-local\"},\"name\":\"antrea\",\"type\":\"antrea\"}" >
I0716 01:28:39.035292     508 server.go:375] Received CmdAdd request cni_args:<container_id:"28bd7d43c70a5984fe5c28803a06d0dc82849474c48e98d2165e52c9d7785c35" netns:"container:d636c4374967a6f90d9b802d9bf24cb4f060e716b0ec8165fcc59be2a59daf28" ifname:"eth0" args:"IgnoreUnknown=1;K8S_POD_NAMESPACE=network-policy-1573;K8S_POD_NAME=server-g4kvm;K8S_POD_INFRA_CONTAINER_ID=28bd7d43c70a5984fe5c28803a06d0dc82849474c48e98d2165e52c9d7785c35" path:"/opt/cni/bin" network_configuration:"{\"capabilities\":{\"dns\":true},\"cniVersion\":\"0.3.0\",\"ipam\":{\"type\":\"host-local\"},\"name\":\"antrea\",\"type\":\"antrea\"}" >
I0716 01:28:39.194301     508 server.go:375] Received CmdAdd request cni_args:<container_id:"d636c4374967a6f90d9b802d9bf24cb4f060e716b0ec8165fcc59be2a59daf28" netns:"none" ifname:"eth0" args:"IgnoreUnknown=1;K8S_POD_NAMESPACE=network-policy-1573;K8S_POD_NAME=server-g4kvm;K8S_POD_INFRA_CONTAINER_ID=d636c4374967a6f90d9b802d9bf24cb4f060e716b0ec8165fcc59be2a59daf28" path:"/opt/cni/bin" network_configuration:"{\"capabilities\":{\"dns\":true},\"cniVersion\":\"0.3.0\",\"ipam\":{\"type\":\"host-local\"},\"name\":\"antrea\",\"type\":\"antrea\"}" >
I0716 01:28:42.341654     508 server.go:375] Received CmdAdd request cni_args:<container_id:"f335e78be911209c7057e1a5fd0a14f9729a203c1be87f0fa274f39d4304f01f" netns:"container:d636c4374967a6f90d9b802d9bf24cb4f060e716b0ec8165fcc59be2a59daf28" ifname:"eth0" args:"IgnoreUnknown=1;K8S_POD_NAMESPACE=network-policy-1573;K8S_POD_NAME=server-g4kvm;K8S_POD_INFRA_CONTAINER_ID=f335e78be911209c7057e1a5fd0a14f9729a203c1be87f0fa274f39d4304f01f" path:"/opt/cni/bin" network_configuration:"{\"capabilities\":{\"dns\":true},\"cniVersion\":\"0.3.0\",\"ipam\":{\"type\":\"host-local\"},\"name\":\"antrea\",\"type\":\"antrea\"}" >
I0716 01:28:42.473665     508 server.go:375] Received CmdAdd request cni_args:<container_id:"28bd7d43c70a5984fe5c28803a06d0dc82849474c48e98d2165e52c9d7785c35" netns:"container:d636c4374967a6f90d9b802d9bf24cb4f060e716b0ec8165fcc59be2a59daf28" ifname:"eth0" args:"IgnoreUnknown=1;K8S_POD_NAMESPACE=network-policy-1573;K8S_POD_NAME=server-g4kvm;K8S_POD_INFRA_CONTAINER_ID=28bd7d43c70a5984fe5c28803a06d0dc82849474c48e98d2165e52c9d7785c35" path:"/opt/cni/bin" network_configuration:"{\"capabilities\":{\"dns\":true},\"cniVersion\":\"0.3.0\",\"ipam\":{\"type\":\"host-local\"},\"name\":\"antrea\",\"type\":\"antrea\"}" >
I0716 01:28:42.514667     508 server.go:375] Received CmdAdd request cni_args:<container_id:"d636c4374967a6f90d9b802d9bf24cb4f060e716b0ec8165fcc59be2a59daf28" netns:"none" ifname:"eth0" args:"IgnoreUnknown=1;K8S_POD_NAMESPACE=network-policy-1573;K8S_POD_NAME=server-g4kvm;K8S_POD_INFRA_CONTAINER_ID=d636c4374967a6f90d9b802d9bf24cb4f060e716b0ec8165fcc59be2a59daf28" path:"/opt/cni/bin" network_configuration:"{\"capabilities\":{\"dns\":true},\"cniVersion\":\"0.3.0\",\"ipam\":{\"type\":\"host-local\"},\"name\":\"antrea\",\"type\":\"antrea\"}" >
```
I think the calls come from https://github.com/kubernetes/kubernetes/blob/5acefcc2bc68b527f22c70f476ca6da52d04c03c/pkg/kubelet/dockershim/helpers_windows.go#L109: whenever `PodSandboxStatus` is called on windows, each container of the pod (including the sandbox) will be used as the parameter of the CNI Add call to get the Pod IP. In above example, `PodSandboxStatus` was called 3 times, the first time usd only the sandbox itself, the second time used the first container and the sandbox (after the first container was created), the third time used two containers and the sandbox (after the second container was created).

And for containerd, I don't have an environment to check, but according to the code, `PodSandboxStatus` wouldn't call CNI Add to fetch pod IP but used the cached return value from the first CNI Add call: 
https://github.com/containerd/containerd/blob/e830c5364cd652b2774dccce3524c7d10f8007dd/pkg/cri/server/sandbox_run.go#L369
https://github.com/containerd/containerd/blob/e830c5364cd652b2774dccce3524c7d10f8007dd/pkg/cri/server/sandbox_status.go#L70-L86

So to correct the docs, maybe we should only replace "kubelet" with "dockershim"?

Reply:
Sure I also wondered why you it was
Referenced as kubelet here as well :)